bin/ch-run.c

```diff
@@ -176,14 +178,74 @@ void env_delta_append(struct env_delta **ds, enum env_action act, char *arg)
 {
    int i;
 
-   for (i = 0; (*ds)[i].action != END; i++) // count existing
+   for (i = 0; (*ds)[i].action != ENV_END; i++) // count existing
          ;
    T_ (*ds = realloc(*ds, (i+2) * sizeof(struct env_delta)));
-   (*ds)[i+1].action = END;
+   (*ds)[i+1].action = ENV_END;
    (*ds)[i].action = act;
    (*ds)[i].arg = arg;
 }
 
+/* Set environment variables as specified in the array lines, which has length
+   line_ct. filename is the source filename, or NULL if source was not a file;
+   in that case, omit line number from any error messages. If expand, then
+   expand variable notation as described in the man page. */
+void envs_set(char **lines, const int line_ct, const char *filename,
+              const bool expand)
+{
+   char *name, *value_old, *value_new, *lineno_str, *item;
+
+   for (int i = 0; i < line_ct; i++) {
+      bool first_written;
+
+      // Skip blank lines.
+      if (lines[i][0] == 0 || lines[i][0] == '\n')
+         continue;
+
+      // Split line into variable name and value.
+      split(&name, &value_old, lines[i], '=');
+      if (filename == NULL)
+         lineno_str = "";
+      else
+         T_ (1 <= asprintf(&lineno_str, ":%d", i+1));
+      Te (name != NULL, "--set-env: no delimiter: %s%s", filename, lineno_str);
+      Te (name[0] != 0, "--set-env: empty name: %s%s", filename, lineno_str);
+
+      // Strip leading and trailing single quotes from value, if both present.
+      if (   strlen(value_old) >= 2
+          && value_old[0] == '\''
+          && value_old[strlen(value_old) - 1] == '\'') {
+         value_old[strlen(value_old) - 1] = 0;
+         value_old++;
+      }
+
+      // Walk through value fragments separated by colon and expand variables
+      // per documentation.
+      value_new = "";
+      first_written = false;
+      item = strsep(&value_old, ":");              // cannot be NULL
+      while (item != NULL) {                       // loop executes ≥ once
+         if (   expand                             // expansion requested
+             && item[0] == '$' && item[1] != 0) {  // ≥1 char in variable name
+            item = getenv(++item);                 // NULL if unset
+            if (item != NULL && item[0] == 0)
+               item = NULL;                        // convert empty to unset
+         }
+         if (item != NULL) {                       // NULL -> omit from output
+            if (first_written)
+               value_new = cat(value_new, ":");
```

Comment:
I reversed the sense here, to prepend colons if not the first item, rather than append if not the last item. When we last discussed this, the two approaches seemed roughly equivalent. But, I realized that we can reliably tell if we've written the first item, but not if an item is the last one (assuming that unset variables expand to nothing rather than the empty string).